crates/bevy_ui/src/focus.rs

```diff
@@ -140,23 +135,88 @@ pub fn ui_focus_system(
-                        state.entities_to_reset.push(entity);
-                    }
-                }
-            } else if *interaction == Interaction::None {
-                *interaction = Interaction::Hovered;
+    set_top_nodes_as_clicked_or_hovered(
+        moused_over_z_sorted_nodes,
+        mouse_button_input,
+        touches_input,
+        state,
+    )
+}
+
+fn contains_cursor(
+    cursor_position: &Vec2,
+    node: &Node,
+    position: Vec3,
+    clip: Option<&CalculatedClip>,
+) -> bool {
+    let ui_position = position.truncate();
+    let extents = node.size / 2.0;
+    let mut min = ui_position - extents;
+    let mut max = ui_position + extents;
+    if let Some(clip) = clip {
+        min = Vec2::max(min, clip.clip.min);
+        max = Vec2::min(max, clip.clip.max);
+    }
+
+    (min.x..max.x).contains(&cursor_position.x) && (min.y..max.y).contains(&cursor_position.y)
+}
+
+fn reset_interactions(
+    node_query: &mut Query<(
+        Entity,
+        &Node,
+        &GlobalTransform,
+        &mut Interaction,
+        Option<&FocusPolicy>,
+        Option<&CalculatedClip>,
+    )>,
+    mouse_button_input: &Input<MouseButton>,
+    touches_input: &Touches,
+    cursor_position: &Option<Vec2>,
+    state: &mut State,
+) {
+    let mouse_release =
+        mouse_button_input.just_released(MouseButton::Left) || touches_input.just_released(0);
+    let input_should_leave_button_clicked = cursor_position.is_some() && !mouse_release;
+
+    for (_entity, _node, _global_transform, mut interaction, _focus_policy, _clip) in
+        node_query.iter_mut()
+    {
+        if input_should_leave_button_clicked && *interaction == Interaction::Clicked {
+            continue;
+        }
+        *interaction = Interaction::None;
+    }
+
+    // reset entities that were both clicked and released in the last frame
+    for entity in state.entities_to_reset.drain(..) {
+        if let Ok(mut interaction) = node_query.get_component_mut::<Interaction>(entity) {
+            *interaction = Interaction::None;
+        }
+    }
+}
+
+fn set_top_nodes_as_clicked_or_hovered(
+    moused_over_z_sorted_nodes: Vec<(Entity, Option<&FocusPolicy>, Mut<Interaction>, FloatOrd)>,
+    mouse_button_input: Res<Input<MouseButton>>,
+    touches_input: Res<Touches>,
+    mut state: Local<State>,
+) {
+    let mouse_released =
+        mouse_button_input.just_released(MouseButton::Left) || touches_input.just_released(0);
+
+    let mouse_clicked =
+        mouse_button_input.just_pressed(MouseButton::Left) || touches_input.just_released(0);
+
+    for (entity, focus_policy, mut interaction, _) in moused_over_z_sorted_nodes {
+        if mouse_clicked {
+            // if the mouse was simultaneously released, reset this Interaction in the next frame
+            if *interaction != Interaction::Clicked && mouse_released {
+                state.entities_to_reset.push(entity);
             }
+            *interaction = Interaction::Clicked;
```

Comment:
- [ ] I might need to be careful here concerning change detection. Double check.

Reply:
I guess after adding a test that checks change detection, unfortunately most of my changes won't work.

At least there are a couple more tests. The `Option<Interaction>` change can be kept, and maybe a refactoring of cutting it into smaller functions. But the approach "set everything to `None`, then check the state" won't work.